src/importlinter/contracts/forbidden.py

```diff
@@ -53,6 +52,10 @@ def check(self, graph: ImportGraph) -> ContractCheck:
                     importer=source_module.name, imported=forbidden_module.name
                 )
                 if chains:
+                    if str(self.allow_indirect_imports).lower() == "true":
```

Comment:
If we're allowing indirect imports, we can actually significantly speed up the algorithm here. (`find_shortest_chains` is potentially extremely expensive.)

Instead of finding all the shortest chains and then removing all the indirect ones, it would be better to use `graph.get_import_details` instead of `find_shortest_chains`. That will just list the direct imports, and is much, much faster.

Also - the `check` method is getting a bit difficult to read. Ideally we would move as much as possible into private methods to keep the `check` method from growing too long.

Reply:
Perhaps, but that's out of scope of this PR. Using graph import details makes this is a 1 line change.